File: web/tests/reference/bst/delete-node-in-a-bst.py
def delete_node(root, key):
    if root is None:
        return None
    if key < root.val:
        root.left = delete_node(root.left, key)
    elif key > root.val:
        root.right = delete_node(root.right, key)
    else:
        if root.left is None:
            return root.right
        if root.right is None:
            return root.left
        successor = root.right
        while successor.left:
            successor = successor.left
        root.val = successor.val
        root.right = delete_node(root.right, successor.val)
    return root

File: web/tests/reference/bst/delete-node-in-a-bst.js
// Promotes the inorder PREDECESSOR where the Python reference promotes the
// successor — both are correct, which is exactly why this lesson is graded
// by a property instead of by one expected tree.
function deleteNode(root, key) {
  if (root === null) return null;
  if (key < root.val) {
    root.left = deleteNode(root.left, key);
  } else if (key > root.val) {
    root.right = deleteNode(root.right, key);
  } else {
    if (root.left === null) return root.right;
    if (root.right === null) return root.left;
    let predecessor = root.left;
    while (predecessor.right) predecessor = predecessor.right;
    root.val = predecessor.val;
    root.left = deleteNode(root.left, predecessor.val);
  }
  return root;
}
